wdl-lint/src/rules/command_shellcheck.rs

```diff
@@ -0,0 +1,357 @@
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that command blocks are free of ShellCheck violations."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "ShellCheck is a static analysis tool and linter for sh / bash. \
+        The lints provided by ShellCheck help prevent common errors and \
+        pitfalls in your scripts. Following its recommendations will increase \
+        the robustness of your command blocks."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Correctness, Tag::Style, Tag::Portability])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::TaskDefinitionNode,
+            SyntaxKind::CommandSectionNode,
+        ])
+    }
+}
+
+/// Convert a WDL placeholder to a bash variable
+fn to_bash_var(placeholder: &Placeholder) -> String {
+    let placeholder_text = placeholder.syntax().text().to_string();
+    let mut bash_var = String::from("_WL");
+    bash_var.push_str(
+        &placeholder_text
+            .chars()
+            .map(|c| {
+                if c.is_ascii_alphanumeric() || c == '_' {
+                    c
+                } else {
+                    '_'
+                }
+            })
+            .collect::<String>(),
+    );
+    bash_var
+}
+
+/// retrieve all input and private declations for a task
+fn task_declarations(task: &TaskDefinition) -> HashSet<String> {
+    let mut decls = HashSet::new();
+    if let Some(input) = task.input() {
+        for decl in input.declarations() {
+            decls.insert(decl.name().as_str().to_owned());
+        }
+    }
+
+    for decl in task.declarations() {
+        decls.insert(decl.name().as_str().to_owned());
+    }
+    decls
+}
+
+/// Creates a "ShellCheck lint" diagnostic from a ShellCheckComment
+fn shellcheck_lint(comment: &ShellCheckComment, span: Span) -> Diagnostic {
+    dbg!(
+        Diagnostic::note("shellcheck violations within a command")
+            .with_rule(ID)
+            .with_label(
+                format!("SC{}[{}]: {}", comment.code, comment.level, comment.message),
+                span,
+            )
+            .with_fix("address ShellCheck violations")
+    )
+}
+
+impl Visitor for CommandSectionShellCheck {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn command_section(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        section: &CommandSection,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        if !program_exists(SHELLCHECK_BIN) {
```

Comment:
Performance nit: this should be done once instead of at every command block.

Reply:
I'm not sure this should be the behavior as well. I made a comment earlier (though it is further down in the code), but I think we should indicate to the user that they need to install `shellcheck` rather than silently return here.

Reply:
We can use a OnceLock here to check if it exists only once and also emit a diagnostic (once) that shellcheck needs to b e installed.